src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -767,25 +768,54 @@ public ISymbolNode DelegateCtor(TypeDesc delegateType, MethodDesc targetMethod,
             TypeAndMethod ctorKey = new TypeAndMethod(delegateType, targetMethod);
             if (!_delegateCtors.TryGetValue(ctorKey, out ctorNode))
             {
-                ModuleToken delegateTypeToken;
-                if (CompilationModuleGroup.ContainsType(delegateType) && delegateType is EcmaType ecmaType)
-                {
-                    delegateTypeToken = new ModuleToken(ecmaType.EcmaModule, (mdToken)MetadataTokens.GetToken(ecmaType.Handle));
-                }
-                else
-                {
-                    // TODO: reverse typedef lookup within the version bubble
-                    throw new NotImplementedException();
-                }
-
                 IMethodNode targetMethodNode = MethodEntrypoint(targetMethod, methodToken, isUnboxingStub: false);
 
                 ctorNode = new DelayLoadHelperImport(this,
                     ILCompiler.DependencyAnalysis.ReadyToRun.ReadyToRunHelper.READYTORUN_HELPER_DelayLoad_Helper,
-                    new DelegateCtorSignature(this, delegateType, delegateTypeToken, targetMethodNode, methodToken));
+                    new DelegateCtorSignature(this, delegateType, default(ModuleToken), targetMethodNode, methodToken));
                 _delegateCtors.Add(ctorKey, ctorNode);
             }
             return ctorNode;
         }
+
+        /// <summary>
+        /// Reverse lookup table mapping external types to reference tokens in the input modules. The table
+        /// gets lazily initialized with all input module references the first time the reverse lookup is
+        /// needed.
+        /// </summary>
+        private Dictionary<TypeDesc, ModuleToken> _typeToRefTokens = null;
+
+        public ModuleToken GetModuleTokenForType(EcmaType type)
+        {
+            if (CompilationModuleGroup.ContainsType(type))
+            {
+                return new ModuleToken(type.EcmaModule, (mdToken)MetadataTokens.GetToken(type.Handle));
+            }
+
+            if (_typeToRefTokens == null)
+            {
+                _typeToRefTokens = new Dictionary<TypeDesc, ModuleToken>();
+                foreach (KeyValuePair<string, string> namePathPair in TypeSystemContext.InputFilePaths)
+                {
+                    EcmaModule module = TypeSystemContext.GetModuleFromPath(namePathPair.Value);
+                    PEReader peReader = module.PEReader;
+                    MetadataReader metadataReader = peReader.GetMetadataReader();
+                    foreach (TypeReferenceHandle typeRefHandle in metadataReader.TypeReferences)
+                    {
+                        TypeDesc resolvedType = (TypeDesc)module.GetObject((EntityHandle)typeRefHandle);
+                        _typeToRefTokens[resolvedType] = new ModuleToken(module, (mdToken)MetadataTokens.GetToken(typeRefHandle));
```

Comment:
Thinking ahead to tests that will compare r2rdump output from crossgen and ILCompiler, can we ensure we'll pick the same answer in situation where multiple assemblies in the input file set have typerefs to the same type? Here, it's defined by the order of assemblies passed in to ILCompiler. Ideally it would be stabilized by, for example, assembly import ordering in the main input assembly.

Reply:
Agreed. Poking around CoreCLR, it seems to me that right now Crossgen doesn't actually have any provisions like this as it likely only supports single-file version bubble in R2R compilation mode. I think we'll need to add at least some rudimentary provisions for multi-module bubble to the legacy CoreCLR Crossgen as according to JanK it should be fairly straightforward and it's the easiest way to obtain code comparisons we can use to monitor CQ.